### PR TITLE
Add Claim Field Validation

### PR DESCRIPTION
**Pull request by syke99**
While this library does provide methods for checking the signature (such as HMCACheck(\<token\>, \<secret\>)) and for checking that the token hasn't expired (*Claims.Valid(time.Time)), it doesn't provide validators for validating each claim in the JWT payload. This PR adds the ability to create a validator for each field to then be used along with the Claims in `ValidatePayloadClaims`. 

As an example:
```go

// Create your field validators. For all minus the TimeFieldValidator, just pass in the expected
// results. For the TimeFieldValidator, pass in the time that all time fields must be valid for. (Does
// not have to be time.Now())
issValidator := jwt.IssuerValidator("testIssuer")
audValidator := jwt.AudiencesValidator([]string{"testAudienceOne", "testAudienceTwo"})
subValidator := jwt.SubjectValidator(strconv.Itoa(1234))
timeValidator := jwt.TimeFieldValidator(time.Now())
idValidator := jwt.IdValidator(strconv.Itoa(5678))

// Not only are there validators for the Registered claims, but there is also a validator for custom claims.
// Just pass in the expected value, and the name of the custom claim field.
customFieldValidator  := jwt.CustomClaimValidator("expectedCustomFieldValue", "customClaimFieldName")

// Pass in the token's claims and your field validators. tokenClaims is *Claims from the created token
err = jwt.ValidatePayloadClaims(tokenClaims, issValidator, audValidator, subValidator, timeValidator, idValidator, customFieldValidator)
```

### CODE REVIEW

**Comment by pascaldekloe** [comment]
Hi Quinn,

Sorry about the late response. I was on Holliday, and I forgot all about this pending request after returning back home. 🙉

Validating the issuer, the subject or a JWT ID is a simple string comparison `==`. Audience has the `AcceptAudience` and timing is covered with `AcceptTemporal`. What this change does is to automate such one-liner checks. An English error string on mismatch may fit some but not all use cases. Consider the following example.

```
if claims.Issuer != MyAuthority {
        // Some issuer managed to produce a token matching our recognised keys.
        // This is either a configuration problem or a serious security breach!
        log.Printf("token from %s validated with an unknown issuer: %s",
                req.RemoteAddr, token)
        SuspiciousTokenCount.Add(1) // update metric counter
        // Don't tell any details to the client. 😬 
        http.Error(res, "token denied", http.StatusUnauthorized)
        return
}

if !claims.AcceptAudience(ThisAppsAudience) {
        TokenHupCount.Add(1)
        reauthWithAudience(res, req, ThisAppsAudience)
        return
}
```

Depending on which part of the validation fails, people may want to log a message (perhaps including the JWT itself), update metrics, provide detailed messages to the user (in any language), etc..

Personally, I really want programmers to write out the validation logic in full, such as shown in the example above. Fluent APIs may look nice, and they probably save a few lines of code, which can be a fit for logic that gets repeated a lot. For security it is more important that you express the checks unambiguously in **one** readable sequence. This code will have to be audited. Changes ought to be rare. The decision tree usually ends up with a few edge cases. I try to make users of this library think about each and everyone of them.

Can you see my point? If you still feel like these helpers have a purpose, then you could make them available in a separate (GitHub) project. I will link to you from the README if wanted. Something like `jwtvalidator.MatchIssuer(string) func(*jwt.Claims) error` perhaps?

File: claim_validate.go [hunk 1]
```diff
@@ -0,0 +1,113 @@
+package jwt
+
+import (
+	"errors"
+	"sort"
+	"time"
+)
+
+type JwtPayloadClaimValidator func(tokenClaims *Claims) error
```
Comment (pascaldekloe) [inline claim_validate.go:9]: Don't repeat yourself in the package name [`jwt.JwtPayloadClaimValidator`]. Don't lowercase abbreviations [`Jwt`]. Do `jwt.Validator` perhaps?

File: claim_validate.go [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+
+func IssuerValidator(expectedIss string) JwtPayloadClaimValidator {
```
Comment (pascaldekloe) [inline claim_validate.go:11]: Document what this validator does, i.e. a case-sensitive match of the string.

File: claim_validate.go [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+	return func(tokenClaims *Claims) error {
+		tokenIss, ok := tokenClaims.String("iss")
+		if !ok {
```
Comment (pascaldekloe) [inline claim_validate.go:14]: Why is it required? Without this check you can match against the empty string (`""`) to require absence of the issuer claim. ;-)

**Comment by syke99** [comment]
No worries!! But following your examples, I do indeed see the logic and reasoning behind it. I also like the idea of more granular errors per validator. I'll get back to you on whether nor not I'll make them available via my own GH project, though!!

File: claim_validate.go [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+		}
+
+		return nil
+	}
+}
+
+func AudiencesValidator(expectedAud []string) JwtPayloadClaimValidator {
+	return func(tokenClaims *Claims) error {
+		tokenAud := tokenClaims.Audiences
+		if len(tokenAud) == 0 {
```
Comment (pascaldekloe) [inline claim_validate.go:44]: Tokens define which audience is required. Absence of such claim implies that all audiences are acceptable. The `AcceptAudience` method follows the RFC specification.

File: claim_validate.go [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+			return errors.New("missing audience claim")
+		}
+
+		if len(tokenAud) != len(expectedAud) {
```
Comment (pascaldekloe) [inline claim_validate.go:48]: Tokens may require less audiences, not more.

File: claim_validate.go [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+			return errors.New("invalid audience claim")
+		}
+
+		sort.Strings(tokenAud)
```
Comment (pascaldekloe) [inline claim_validate.go:52]: Faster to just loop than an extra sort operation.

File: claim_validate.go [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+		}
+
+		return nil
+	}
+}
+
+func TimeFieldValidator(expectedTime time.Time) JwtPayloadClaimValidator {
+	return func(tokenClaims *Claims) error {
+		if ok := tokenClaims.Valid(expectedTime); !ok {
+			return errors.New("token has expired")
```
Comment (pascaldekloe) [inline claim_validate.go:68]: It could be that the not-before was in the future, which is something else than expiry.

File: claim_validate.go [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+		}
+
+		return nil
+	}
+}
+
+func IdValidator(expectedId string) JwtPayloadClaimValidator {
+	return func(tokenClaims *Claims) error {
+		tokenId, ok := tokenClaims.String("jti")
```
Comment (pascaldekloe) [inline claim_validate.go:77]: The `jti` is in the `ID` field from `tokenClaims` [`jwt.Registered`].

File: claim_validate.go [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+
+		if tokenId != expectedId {
+			return errors.New("invalid id claim")
+		}
+
+		return nil
+	}
+}
+
+func CustomClaimValidator(expectedValue, customField string) JwtPayloadClaimValidator {
```
Comment (pascaldekloe) [inline claim_validate.go:90]: `StringClaimValidator` as it leaves room for a number check and such...

File: claim_validate_test.go [hunk 10]
```diff
@@ -0,0 +1,705 @@
...
+	Validator Creations
+*/
+
+// Test IssValidator function, it is not expected to return an error
+func TestIssValidator(t *testing.T) {
+	// Act
+	validator := IssuerValidator("testing")
+
+	// Assert
+	if validator == nil {
```
Comment (pascaldekloe) [inline claim_validate_test.go:21]: Not really a useful check. Got to validate the happy flow plus each error scenario.

**Pull request closed without merging**